js-client/src/app-main.ts

```diff
@@ -0,0 +1,95 @@
+import { LitElement, css, html } from "lit";
+import { customElement, queryAsync, state } from "lit/decorators.js";
+import "@shoelace-style/shoelace/dist/themes/light.css";
+import type { ClickItemPayload } from "./nix-search";
+import "./nix-search.ts";
+
+import dagre from "cytoscape-dagre";
+import cytoscape from "cytoscape";
+
+const API_URL = __API_URL__;
+cytoscape.use(dagre);
+
+async function getGraph(pkgName: string) {
+  const response = await fetch(`${API_URL}/gremlin`, {
+    method: "POST",
+    headers: {
+      Accept: "application/json",
+      "Content-Type": "application/json",
+    },
+    body: JSON.stringify({
+      query: `g.V().filter{it.get().value('pname').matches('${pkgName}')}.repeat(outE().otherV().simplePath()).until(__.not(outE().simplePath())).path().by('pname').by(label)`,
+    }),
+  });
+  return await response.json();
+}
+
+function renderCyGraph(graphData: any, container: HTMLElement) {
+  console.log(graphData);
+  const data = graphData.cyto["graph-data"];
+  cytoscape({
+    container,
+    layout: { name: "dagre" },
+    ...data,
+    style: [
+      {
+        selector: "node",
+        style: {
+          label: "data(id)",
+        },
+      },
+    ],
+  });
+}
+
+@customElement("app-main")
+export class AppMain extends LitElement {
+  @queryAsync("#cy") _cy: Promise<HTMLElement>;
+  @state() _error = false;
+  render() {
+    return html`
+      <nix-search @click-item=${this.updateGraph}> </nix-search>
+      <div id="cy-container">
+        ${this._error
+          ? html`<div>Error fetching graph data</div>`
+          : html`<div id="cy"></div>`}
+      </div>
+    `;
+  }
+
+  private async updateGraph(ev: CustomEvent<ClickItemPayload>) {
+    try {
+      this._error = false; // Remove old errors
+      const cy = await this._cy;
+      const graphData = await getGraph(ev.detail.name);
+      renderCyGraph(graphData, cy);
+    } catch {
+      this._error = true;
+    }
+  }
```

Comment:
It seems confusing to me to have a method `updateGraph` which is an event handler which is passed to `nix-search`, instead of being bound to the `AppMain` it belongs.

Would it be possible

to have a `interface UpdateGraphRequestPayload { name: string; }` type in `app-main.ts`

then to have an event handler in `AppMain`

```suggestion
  private async handleUpdateGraphRequest(ev: CustomEvent<UpdateGraphRequestPayload>) {
      // will be called when a request to update the graph view is dispatched
```

then to have the logic to listen to such events stay in `AppMain` and not leak to other components?

Then other components like `NixSearch` can dispatch events without being tightly coupled.

Reply:
`app-main` is not passing anything to `nix-search`. `nix-search` is not coupled to `app-main`, it doesn't know anything about the parent component. When the user click on one of the packages, it emits a custom event with the package's name. Parent elements can decide if listen or ignore the event. 

Maybe there is some confusion because I called two different functions `updateGraph` in `app-main` and in `nix-search`: 
https://github.com/tweag/nixpkgs-graph-explorer/blob/9847650718357b20bb28987a7b6cc8ca5ff8f70b/js-client/src/nix-search.ts#L94
Those functions are independent, and thinking again about it, I should rename the function in `nix-search`

Once the logic is more complex, we should probably move the update graph logic to its own component.